cluster/cluster.go

```diff
@@ -171,26 +171,27 @@ func GetLocalKubeConfig(configPath string) string {
 }
 
 func rebuildLocalAdminConfig(kubeCluster *Cluster) error {
-	logrus.Infof("[reconcile] Rebuilding and update local kube config")
-	var workingConfig string
+	logrus.Infof("[reconcile] Rebuilding and updating local kube config")
+	var workingConfig, newConfig string
 	currentKubeConfig := kubeCluster.Certificates[pki.KubeAdminCommonName]
 	caCrt := kubeCluster.Certificates[pki.CACertName].Certificate
 	for _, cpHost := range kubeCluster.ControlPlaneHosts {
-		newConfig := pki.GetKubeConfigX509WithData(
-			"https://"+cpHost.Address+":6443",
-			pki.KubeAdminCommonName,
-			string(cert.EncodeCertPEM(caCrt)),
-			string(cert.EncodeCertPEM(currentKubeConfig.Certificate)),
-			string(cert.EncodePrivateKeyPEM(currentKubeConfig.Key)))
-
+		if (currentKubeConfig == pki.CertificatePKI{}) {
+			kubeCluster.Certificates = make(map[string]pki.CertificatePKI)
+			newConfig = getLocalAdminConfigWithNewAddress(kubeCluster.LocalKubeConfigPath, cpHost.Address)
+		} else {
+			kubeURL := fmt.Sprintf("https://%s:6443", cpHost.Address)
+			caData := string(cert.EncodeCertPEM(caCrt))
+			crtData := string(cert.EncodeCertPEM(currentKubeConfig.Certificate))
+			keyData := string(cert.EncodePrivateKeyPEM(currentKubeConfig.Key))
+			newConfig = pki.GetKubeConfigX509WithData(kubeURL, pki.KubeAdminCommonName, caData, crtData, keyData)
+		}
 		if err := pki.DeployAdminConfig(newConfig, kubeCluster.LocalKubeConfigPath); err != nil {
 			return fmt.Errorf("Failed to redeploy local admin config with new host")
 		}
 		workingConfig = newConfig
-		if _, err := GetK8sVersion(kubeCluster.LocalKubeConfigPath); err != nil {
-			logrus.Infof("[reconcile] host [%s] is not active master on the cluster", cpHost.Address)
-			continue
-		} else {
+		if _, err := GetK8sVersion(kubeCluster.LocalKubeConfigPath); err == nil {
+			logrus.Infof("[reconcile] host [%s] is active master on the cluster", cpHost.Address)
```

Comment:
shall we move `workingConfig = newConfig` inside the if condition? otherwise it would be set even when the getk8sversion call fails

Reply:
Actually that has been done on purpose, because there are cases where we need a new config with even a not working control plane host, consider this scenario:

- run rke with one control hosts
- remove these old control host and add new one in the cluster.yml
- run rke up again

rke in that case will pull the certificates from the old node and rebuild local admin with the new (and not working) host and use it after it deploys all components to save kubernetes certificates in the end